contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,853 @@
+target state, and it therefor interacts well with canaries and staged roll outs.
+Note that, while users can update the target template revision, they can not 
+update the current template revision. The only way to advance the current 
+template revision is to successfully complete an update.
+
+## Tests
+- Updating a StatefulSet's containers will trigger updates to the StatefulSet's 
+Pods respecting the 
+[identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity) 
+and [deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+guarantees.
+- A StatefulSet update will block on failure.
+- A StatefulSet update can be rolled back.
+- A StatefulSet update can be rolled forward by applying another update. 
+- A StatefulSet update's status can be retrieved.
+- A StatefulSet's revision history contains all updates with respect to the 
+configured revision history limit.
+- A StatefulSet update can create a canary.
+- A StatefulSet update can be performed in stages.
+
+## Future Work
+In the future, we may implement the following features to enhance StatefulSet 
+updates.
+
+### Termination Reason
+Without communicating a signal indicating the reason for termination to a Pod in 
+a StatefulSet, as proposed [here](https://github.com/kubernetes/kubernetes/issues/1462),
+the tenant application has no way to determine if it is being terminated due to 
+a scale down operation or due to an update. 
+Consider a BASE distributed storage application like Cassandra, where 2 TiB of 
+persistent data is not atypical, and the data distribution is not identical on 
+every server. We want to enable two distinct behaviors based on the reason for 
+termination.
+
+- If the termination is due to scale down, during the configured termination 
+grace period, the entry point of the Pod should cause the application to drain 
+its client connections, replicate its persisted data (so that the cluster is not 
+left under replicated) and decommission the application to remove it from the 
+cluster.
+- If the termination is due to a temporary capacity loss (e.g. an update or an 
+image upgrade), the application should drain all of its client connections, 
+flush any in memory data structures to the file system, and synchronize the 
+file system with storage media. It should not redistribute its data.
+
+If the application implements the strategy of always redistributing its data, 
+we unnecessarily decrease recovery time during an update and incur the 
+additional network and storage cost of two full data redistributions for every 
+updated node.
+It should be noted that this is already an issue for Node cordon and Pod eviction 
+(due to drain or taints), and applications can use the same mitigation as they 
+would for these events for StatefulSet update.
+
+### VolumeTemplatesSpec Updates
```

Comment:
Wouldn't this also be valuable for the simpler case where your data is growing and you want to vertically scale each stateful pod by increasing its volume size? Or is there some reason you only mention temporary changes as being relevant here?